File: drivers/test-ultrasonic/src/driver.py
"""ultrasonic driver implementation."""

import asyncio
import json
import logging
from datetime import datetime, timezone
from typing import Any, Dict, Optional

import nats
from aiohttp import web
from prometheus_client import Counter, Gauge, generate_latest

from .config import Config
from .hal import HALMessage

logger = logging.getLogger(__name__)


class ultrasonicDriver:
    """ultrasonic sensor driver."""

    def __init__(self, config: Config) -> None:
        self.config = config
        self.nc: Optional[nats.NATS] = None
        self.js: Optional[nats.JetStreamContext] = None
        self._running = False
        
        # Metrics
        self.messages_published = Counter(
            'tafy_driver_messages_published_total',
            'Total messages published',
            ['driver']
        )
        self.messages_failed = Counter(
            'tafy_driver_messages_failed_total', 
            'Total failed messages',
            ['driver']
        )
        self.readings = Gauge(
            'tafy_sensor_ultrasonic_value',
            'Current ultrasonic sensor reading'
        )
        
        # HTTP server
        self.app = web.Application()
        self._setup_routes()

    def _setup_routes(self) -> None:
        """Setup HTTP routes."""
        self.app.router.add_get('/health', self._health_handler)
        self.app.router.add_get('/metrics', self._metrics_handler)

    async def run(self) -> None:
        """Run the driver."""
        # Connect to NATS
        self.nc = await nats.connect(self.config.nats.url)
        self.js = self.nc.jetstream()
        
        # Subscribe to commands
        await self.nc.subscribe(
            f"hal.v1.sensor.ultrasonic.cmd",
            cb=self._handle_command
        )
        
        # Start HTTP server
        runner = web.AppRunner(self.app)
        await runner.setup()
        site = web.TCPSite(runner, '0.0.0.0', 8080)
        asyncio.create_task(site.start())
        
        # Main loop
        self._running = True
        interval = 1.0 / self.config.sample_rate
        
        while self._running:
            try:
                await self._process_loop()
                await asyncio.sleep(interval)
            except Exception as e:
                logger.error(f"Error in main loop: {e}")
                self.messages_failed.labels(driver='ultrasonic').inc()

    
    async def _process_loop(self) -> None:
        """Read sensor and publish data."""
        # TODO: Read from actual hardware
        value = 42.0  # Replace with actual sensor reading
        
        # Update metric
        self.readings.set(value)
        
        # Create HAL message
        message = HALMessage(
            hal_major=1,
            hal_minor=0,
            schema="tafylabs/hal/sensor/ultrasonic/1.0",
            device_id=self.config.device_id,
            caps=["sensor.ultrasonic:v1.0"],
            ts=datetime.now(timezone.utc).isoformat(),
            payload={
                "value": value,
                "unit": "TODO"
            }
        )
        
        # Publish
        await self.nc.publish(
            "hal.v1.sensor.ultrasonic.data",
            json.dumps(message.model_dump()).encode()
        )
        self.messages_published.labels(driver='ultrasonic').inc()
    

    async def _handle_command(self, msg: nats.Msg) -> None:
        """Handle incoming commands."""
        try:
            data = json.loads(msg.data.decode())
            command = HALMessage(**data)
            
            # TODO: Handle commands based on payload
            logger.info(f"Received command: {command.payload}")
            
            # Send acknowledgment
            ack = HALMessage(
                hal_major=1,
                hal_minor=0,
                schema="tafylabs/hal/sensor/ultrasonic/ack/1.0",
                device_id=self.config.device_id,
                caps=["sensor.ultrasonic:v1.0"],
                ts=datetime.now(timezone.utc).isoformat(),
                payload={
                    "status": "ok",
                    "command": command.payload
                }
            )
            
            await self.nc.publish(
                f"hal.v1.sensor.ultrasonic.ack",
                json.dumps(ack.model_dump()).encode()
            )
            
        except Exception as e:
            logger.error(f"Failed to handle command: {e}")

    async def _health_handler(self, request: web.Request) -> web.Response:
        """Health check handler."""
        return web.json_response({
            "status": "healthy",
            "device_id": self.config.device_id,
            "uptime": (datetime.now() - self.config.start_time).total_seconds()
        })

    async def _metrics_handler(self, request: web.Request) -> web.Response:
        """Prometheus metrics handler."""
        metrics = generate_latest()
        return web.Response(
            body=metrics,
            content_type="text/plain; version=0.0.4"
        )

    def stop(self) -> None:
        """Stop the driver."""
        self._running = False

    async def cleanup(self) -> None:
        """Cleanup resources."""
        if self.nc and self.nc.is_connected:
            await self.nc.close()

File: tools/tafy-cli/drivers/test-motor-cpp/src/driver.cpp
#include "driver.hpp"
#include <iostream>
#include <chrono>
#include <thread>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

motorDriver::motorDriver(const Config& config) : config_(config) {
    // Initialize NATS connection
    natsStatus s = natsConnection_ConnectTo(&nc_, config_.nats.url.c_str());
    if (s != NATS_OK) {
        throw std::runtime_error("Failed to connect to NATS: " + std::string(natsStatus_GetText(s)));
    }

    // Setup metrics
    setupMetrics();
    
    // Start HTTP server
    startHttpServer();
}

motorDriver::~motorDriver() {
    stop();
    
    if (cmd_sub_) {
        natsSubscription_Destroy(cmd_sub_);
    }
    
    if (nc_) {
        natsConnection_Close(nc_);
        natsConnection_Destroy(nc_);
    }
}

void motorDriver::run() {
    // Subscribe to commands
    std::string cmd_topic = "hal.v1.actuator.motor.cmd";
    natsStatus s = natsConnection_Subscribe(&cmd_sub_, nc_, cmd_topic.c_str(), 
                                          commandCallback, this);
    if (s != NATS_OK) {
        throw std::runtime_error("Failed to subscribe: " + std::string(natsStatus_GetText(s)));
    }

    running_ = true;
    
    // Main loop
    auto interval = std::chrono::milliseconds(1000 / config_.sample_rate);
    while (running_) {
        processLoop();
        std::this_thread::sleep_for(interval);
    }
}

void motorDriver::stop() {
    running_ = false;
}

void motorDriver::processLoop() {
    try {
        
        // TODO: Implement main processing loop for actuator
        
    } catch (const std::exception& e) {
        std::cerr << "Error in process loop: " << e.what() << std::endl;
        if (messages_failed_) {
            messages_failed_->Increment();
        }
    }
}

void motorDriver::publishData(double value) {
    HALMessage msg;
    msg.hal_major = 1;
    msg.hal_minor = 0;
    msg.schema = "tafylabs/hal/actuator/motor/1.0";
    msg.device_id = config_.device_id;
    msg.caps = {"actuator.motor:v1.0"};
    msg.ts = getCurrentTimestamp();
    
    json payload;
    payload["value"] = value;
    payload["unit"] = "TODO";
    msg.payload = payload.dump();
    
    std::string data = msg.toJson();
    std::string topic = "hal.v1.actuator.motor.data";
    
    natsStatus s = natsConnection_PublishString(nc_, topic.c_str(), data.c_str());
    if (s != NATS_OK) {
        throw std::runtime_error("Failed to publish: " + std::string(natsStatus_GetText(s)));
    }
    
    if (messages_published_) {
        messages_published_->Increment();
    }
}

void motorDriver::handleCommand(natsMsg* msg) {
    try {
        const char* data = natsMsg_GetData(msg);
        int dataLen = natsMsg_GetDataLength(msg);
        
        std::string jsonStr(data, dataLen);
        HALMessage cmd = HALMessage::fromJson(jsonStr);
        
        // TODO: Handle commands based on payload
        std::cout << "Received command: " << cmd.payload << std::endl;
        
        // Send acknowledgment
        HALMessage ack;
        ack.hal_major = 1;
        ack.hal_minor = 0;
        ack.schema = "tafylabs/hal/actuator/motor/ack/1.0";
        ack.device_id = config_.device_id;
        ack.caps = {"actuator.motor:v1.0"};
        ack.ts = getCurrentTimestamp();
        
        json ackPayload;
        ackPayload["status"] = "ok";
        ackPayload["command"] = json::parse(cmd.payload);
        ack.payload = ackPayload.dump();
        
        std::string ackTopic = "hal.v1.actuator.motor.ack";
        natsConnection_PublishString(nc_, ackTopic.c_str(), ack.toJson().c_str());
        
    } catch (const std::exception& e) {
        std::cerr << "Failed to handle command: " << e.what() << std::endl;
    }
    
    natsMsg_Destroy(msg);
}

void motorDriver::commandCallback(natsConnection* nc, natsSubscription* sub, 
                                    natsMsg* msg, void* closure) {
    (void)nc;
    (void)sub;
    
    auto* driver = static_cast<motorDriver*>(closure);
    driver->handleCommand(msg);
}

void motorDriver::setupMetrics() {
    registry_ = std::make_shared<prometheus::Registry>();
    
    auto& messages_published_family = prometheus::BuildCounter()
        .Name("tafy_driver_messages_published_total")
        .Help("Total messages published")
        .Register(*registry_);
    messages_published_ = &messages_published_family.Add({"driver", "motor"});
    
    auto& messages_failed_family = prometheus::BuildCounter()
        .Name("tafy_driver_messages_failed_total")
        .Help("Total failed messages")
        .Register(*registry_);
    messages_failed_ = &messages_failed_family.Add({"driver", "motor"});
    
    
}

void motorDriver::startHttpServer() {
    exposer_ = std::make_unique<prometheus::Exposer>("0.0.0.0:8080");
    exposer_->RegisterCollectable(registry_);
    
    // Note: prometheus-cpp handles /metrics endpoint
    // For /health endpoint, you'd need a separate HTTP server
    // This is simplified for the template
}
